autopilot/prefattach.go

```diff
@@ -61,45 +52,8 @@ var _ AttachmentHeuristic = (*ConstrainedPrefAttachment)(nil)
 func (p *ConstrainedPrefAttachment) NeedMoreChans(channels []Channel,
 	funds btcutil.Amount) (btcutil.Amount, uint32, bool) {
 
-	// If we're already over our maximum allowed number of channels, then
-	// we'll instruct the controller not to create any more channels.
-	if len(channels) >= int(p.chanLimit) {
-		return 0, 0, false
-	}
-
-	// The number of additional channels that should be opened is the
-	// difference between the channel limit, and the number of channels we
-	// already have open.
-	numAdditionalChans := uint32(p.chanLimit) - uint32(len(channels))
-
-	// First, we'll tally up the total amount of funds that are currently
-	// present within the set of active channels.
-	var totalChanAllocation btcutil.Amount
-	for _, channel := range channels {
-		totalChanAllocation += channel.Capacity
-	}
-
-	// With this value known, we'll now compute the total amount of fund
-	// allocated across regular utxo's and channel utxo's.
-	totalFunds := funds + totalChanAllocation
-
-	// Once the total amount has been computed, we then calculate the
-	// fraction of funds currently allocated to channels.
-	fundsFraction := float64(totalChanAllocation) / float64(totalFunds)
-
-	// If this fraction is below our threshold, then we'll return true, to
-	// indicate the controller should call Select to obtain a candidate set
-	// of channels to attempt to open.
-	needMore := fundsFraction < p.threshold
-	if !needMore {
-		return 0, 0, false
-	}
-
-	// Now that we know we need more funds, we'll compute the amount of
-	// additional funds we should allocate towards channels.
-	targetAllocation := btcutil.Amount(float64(totalFunds) * p.threshold)
-	fundsAvailable := targetAllocation - totalChanAllocation
-	return fundsAvailable, numAdditionalChans, true
+	// We'll try to open more channels as long as the constraints allow it.
```

Comment:
So with this change to the scoring based heuristics, I envisioned removing this method `NeedMoreChans` all together. Instead the agent determines this.

Reply:
So with this change, we'd have an interface something like: 
```go
type AttachmentHeuritic interface {
    SelectTopN(self *btcec.PublicKey, graph ChannelGraph, skipNodes map[NodeID]struct{}) ([]AttachmentScore, error) 
}
```

Eventually once we move to the retry/queue based model, we remove the `skipNodes` all together.

Reply:
In this case `AttachmentScore` is more or less a two-tuple, of the node info (pubkey, etc) and the `float64` score (or an int, whichever).